feat: conditional exports

ricokahler opened the pull request:
This PR:
1. Adds conditional exports to color2k (see below)
2. Includes #405 
3. Cleans up some ESLint package deps

Color2k handles its own bundle with the following entry points and [targets](https://github.com/browserslist/browserslist):

- `main` — `node 4`
- `unpkg` — browserslist `defaults`
- `module` — `node 6` for ES2015 support defined by [rollup](https://github.com/rollup/rollup/wiki/pkg.module) a while ago
- `exports.import` — `node 12` for ES2019 support (esbuild and node >12 in ES mode uses this entry point)
- `exports.default` — `node 12` for CJS node 12 support, since export maps weren't support until node 12
- `exports.package.json` — for backwards compat to importing `package.json` (since it's restricted to import other files now)

---

The following consumers have been manually tested:

- [x] legacy node (e.g. node 4)
- [x] `<script type="module">` via skypack
- [x] deno import via skypack
- [x] jest with the default babel setup — this uses jest-resolve which uses [`resolve`](https://github.com/browserify/resolve) (maintained by none other than @ljharb! 😄). This works but uses the legacy `main` CJS bundle due to the complexity of closing https://github.com/browserify/resolve/issues/222. Note from the jest folks, they are [highly unlikely to switch from resolve](https://github.com/facebook/jest/issues/9771#issuecomment-776681032).
- [x] `@babel/register` with `[@babel/preset-env, {targets: 'node 12'}]` — uses the CJS `exports.require` entry point
- [x] `@rollup/plugin-node-resolve` — uses the `exports.import` entry point
- [x] webpack 5 — uses enhanced-resolve with the conditions `["require", "module"]` for CJS and `["import", "module"]` for ESM. resolves to the `exports.import` entry point
- [x] webpack 3 — this works and uses the legacy [`module`](https://github.com/rollup/rollup/wiki/pkg.module) entry point
- [x] node LTS with both ESM/CJS — works as expected

Closes #413

ricokahler committed:
feat: conditional exports

Co-authored-by: ljharb <ljharb@gmail.com>

BREAKING CHANGE: This package now includes [conditional package exports][docs]. Most consumers should be able to upgrade without any issues, however, package exports limits the allowed entry points inside of Node.js and also affects consumption in other bundlers.

- `main` targets been updated to support down to node 4 (ES5)
- `module` targets has update to support down to node 6 (ES2015)
- `exports` targets have been set to node 12 (ES2019)
- `unpkg` targets have been set to browserslist `defaults`

[docs]: https://nodejs.org/api/packages.html#conditional-exports

ljharb commented on package.json:
```diff
@@ -22,14 +22,18 @@
-  "unpkg": "./dist/index.umd.js",
-  "module": "./dist/index.module.js",
-  "source": "./src/index.ts",
+  "main": "./dist/index.main.umd.js",
```
why is the "main" a UMD? you could do that in a "browser" perhaps, but node doesn't need a UMD (and really, neither does anyone else)

ricokahler replied:
The UMD is primarily for raw browser usage for both unpkg and jsdeliver. I think skypack does a better job at kind of use case but I liked the idea of a legacy catch-all UMD bundle.

unpkg:

> If you omit the file path (i.e. use a “bare” URL), unpkg will serve the file specified by the `unpkg` field in package.json, or fall back to `main`.
>
> [[source]](https://unpkg.com/)

jsdelivr:

> For packages hosted on npm, we support setting a "default" file for each package. This file is displayed at the top of directory listings on our website and available under shorter URL on the CDN. The default file can be configured by setting one of the following fields in package.json (ordered by priority):
>
> 1. jsdelivr
> 2. browser
> 3. main
>
> [[source]](https://www.jsdelivr.com/features#publishing-packages)

ljharb replied:
Right - I think it's best not to penalize node users, or modern bundler users, by giving them a UMD. The "unpkg" key seems fine to include a UMD.

ljharb commented on package.json:
```diff
@@ -22,14 +22,18 @@
+    ".": {
+      "import": "./dist/index.exports.import.es.mjs",
+      "require": "./dist/index.exports.require.cjs.js"
+    },
```
```suggestion
    ".": [
      {
        "import": "./dist/index.exports.import.es.mjs",
        "require": "./dist/index.exports.require.cjs.js",
        "default": "./dist/index.exports.require.cjs.js"
      },
      "./dist/index.exports.require.cjs.js"
    ],
```
otherwise it'll break in a few node 13 versions. you could also just omit "require" entirely, but you always need "default".

ricokahler replied:
ah nice. i didn't know it supported an array. i'll give this a try this week

---

>  you could also just omit "require" entirely, but you always need "default".

I'm looking at `@rollup/plugin-node-resolve` and it says:

> By default, this plugin looks for the `['default', 'module', 'import']` conditions when resolving imports.
>
> [[source]](https://github.com/rollup/plugins/tree/9874740ca685b1600e4319365c21e8c1131e9d95/packages/node-resolve#exportconditions)

Do you know if the order matters there? I intentionally left out `default` there because I wanted the rollup plugin to resolve `module` before default

ljharb replied:
the order does matter, yes, so having "default" last is what you want.

Also, "module" isn't a condition, so it's super weird that that rollup would look for it.

ricokahler replied:
I was looking deeper into this and the order of the conditions matched by `@rollup/plugin-node-resolve` does not seem to matter.

I was looking at a matching implementation [here](https://github.com/lukeed/resolve.exports/tree/6d29cfbabfea13a41cbb75b57bdf46f6b0bb5115#optionsconditions) with this nice explainer

> **Important:** The order specified within `options.conditions` does not matter.
> The matching order/priority is **always** determined by the `"exports"` map's key order.

…which made things click for myself.

This may have been what you meant but I was still confused 😅. Anyway I put the `default` back in there with the CJS bundle and removed `require`.

---

> Also, "module" isn't a condition, so it's super weird that that rollup would look for it.

This condition is noted in webpack's [package exports doc](https://webpack.js.org/guides/package-exports/#reference-syntax). It's nice if you wanted to target those bundlers but the output was redundant to the `import` entry so I removed it.

ljharb reviewed (changes requested):
This works in theory :-) i'd have to look at the build output to be sure.

ljharb reviewed (approved):
seems good; short of an actual node 4 test

ricokahler commented:
@ljharb alright i've spent waaaay to long reading up on these exports and manually testing but I think we're finally good to go after a node 4 test. I'll merge after getting that done.

Thanks for your feedback!

ljharb commented:
I will point out that adding "exports" is almost always a breaking change, so you may want to do this as a semver-major bump.

ricokahler committed:
chore: update eslint and config

ricokahler committed:
feat: allow custom bg color in `hasBadContrast`

Co-authored-by: Novolokov <83767977+Novolokov@users.noreply.github.com>

Squashed commit of the following:

commit 32afa52f8030cef609ab8b736ad8a9b4395b8373
Merge: d0b4620 37af5b7
Author: Rico Kahler <ricokahler@gmail.com>
Date:   Sun Feb 13 19:41:05 2022 -0500

    Merge branch 'main' into patch-1

commit d0b46203750592d2ffff1f04fd92c64b293c1bad
Author: Novolokov <83767977+Novolokov@users.noreply.github.com>
Date:   Sat Dec 4 13:04:24 2021 +0100

    allow custom background color

ricokahler committed:
chore: add dry-run to semantic-release
